Android/app/src/main/java/app/intra/ServerChooserFragment.java

```diff
@@ -0,0 +1,155 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package app.intra;
+
+import android.app.Dialog;
+import android.os.Bundle;
+import android.support.v7.app.AlertDialog;
+import android.support.v7.preference.PreferenceDialogFragmentCompat;
+import android.text.Editable;
+import android.text.TextWatcher;
+import android.view.KeyEvent;
+import android.view.View;
+import android.widget.Button;
+import android.widget.EditText;
+import android.widget.RadioGroup;
+import android.widget.TextView;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+
+import app.intra.util.Untemplate;
+
+/**
+ * User interface for a the server URL selection.
+ */
+
+public class ServerChooserFragment extends PreferenceDialogFragmentCompat
+    implements RadioGroup.OnCheckedChangeListener, TextWatcher, EditText.OnEditorActionListener {
+    private String newUrl = null;
+    private RadioGroup buttons = null;
+    private EditText text = null;
+
+    public static ServerChooserFragment newInstance(String key) {
+        final ServerChooserFragment fragment = new ServerChooserFragment();
+        final Bundle bundle = new Bundle(1);
+        bundle.putString(ARG_KEY, key);
+        fragment.setArguments(bundle);
+        return fragment;
+    }
+
+    @Override
+    public void onCheckedChanged(RadioGroup group, int checkedId) {
+        if (checkedId == R.id.pref_server_google) {
+            newUrl = getResources().getString(R.string.url0);
+            setValid(true);
+            text.setEnabled(false);
+        } else if (checkedId == R.id.pref_server_cloudflare) {
+            newUrl = getResources().getString(R.string.url1);
+            setValid(true);
+            text.setEnabled(false);
+        } else {
+            newUrl = text.getText().toString();
+            setValid(checkUrl(Untemplate.strip(newUrl)));
+            text.setEnabled(true);
+        }
+    }
+
+    @Override
+    public void beforeTextChanged(CharSequence s, int start, int count, int after) {
+    }
+
+    @Override
+    public void onTextChanged(CharSequence s, int start, int before, int count) {
+    }
+
+    @Override
+    public void afterTextChanged(Editable s) {
+        newUrl = s.toString();
+        setValid(checkUrl(Untemplate.strip(newUrl)));
+    }
+
+    // Check that the URL is a plausible DOH server: https with a domain, a path (at least "/"),
+    // and no query parameters or fragment.
+    private boolean checkUrl(String url) {
+        try {
+            URL parsed = new URL(url);
+            return parsed.getProtocol().equals("https") && !parsed.getHost().isEmpty() &&
+                !parsed.getPath().isEmpty() && parsed.getQuery() == null && parsed.getRef() == null;
```

Comment:
Do you think we really need to have a path?  People type in google.com [Enter] all the time?

Reply:
The problem is, entering a bare domain name is (1) very tempting and (2) not going to work (unless the server is actually running the API endpoint on `/`, which would be very strange).  Requiring a path at least makes this explicit, and hopefully reduces the frequency of user error.

Reply:
Fair enough